### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by AkshataDudhane**
Flask Assignment

### CODE REVIEW

File: orders_app.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        "created_time": order_model.created_time,
+        "updated_time": order_model.updated_time,
+        "is_delivered": False
+    }
+
+    return jsonify({"message":"order created successfully","order_id":str(order_model.id),"response":response_data}), 201
+
+@app.route('/order/<order_id>', methods=['DELETE'])
+def delete_order(order_id):
+    order = OrdersModel.objects(id=order_id).first()
```
Comment (AkshataDudhane) [inline orders_app.py]: .get

File: orders_app.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+connect('OrderDb')
+
+@app.route('/order', methods=['POST'])
+def create_order():
+    data = request.json
+    
+    # Validate input data using Order class methods
+    order = Order(order_id=id,name=data['name'], birthday=data['birthday'], email=data['email'], state=data['state'], zipcode=data['zipcode']) 
+    if not order.user.check_state():
```
Comment (AkshataDudhane) [inline orders_app.py]: generic validation method

File: orders_app.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+def list_orders():
+    page = int(request.args.get('page', 1))
+    per_page = 10
+    filter_email = request.args.get('email')
+    filter_state = request.args.get('state')
+    filter_zipcode = request.args.get('zipcode')
+    sort_order = int(request.args.get('sort_order', 1))  # 1 for ascending, -1 for descending
+
+    query = OrdersModel.objects(is_deleted=False)
+
```
Comment (AkshataDudhane) [inline orders_app.py]: add validations

File: orders_app.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    if filter_email:
+        query = query.filter(email=filter_email)
+    if filter_state:
+        query = query.filter(state=filter_state)
+    if filter_zipcode:
+        query = query.filter(zipcode=filter_zipcode)
+
+    total_orders = query.count()
+    orders = query.order_by(sort_order * 'created_time')[per_page * (page - 1):per_page * page]
```
Comment (AkshataDudhane) [inline orders_app.py]: add start index and end index

File: orders_app.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        "updated_time": order.updated_time,
+        "is_delivered": order.is_delivered,
+        "is_deleted": order.is_deleted
+        
+    }
+
+    return jsonify({"response":response_data}), 200
+
+
+@app.route('/orders', methods=['GET'])
```
Comment (AkshataDudhane) [inline orders_app.py:77]: need to check this route

File: orders_app.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+}
+), 404
+    
+    order.update(set__is_deleted=True, set__del_time=datetime.utcnow())
+    
+    return jsonify({"is_deleted":bool(order.is_deleted),"message": "Order marked as deleted"}), 200
+
+@app.route('/order/<order_id>', methods=['GET']) #get order by id
+def get_order(order_id):
+    order = OrdersModel.objects(id=order_id).first()
```
Comment (AkshataDudhane) [inline orders_app.py]: add filter for deleted fields

File: orders_app.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        return jsonify({
+       "error_code" : "404",
+       "error" : "ORDER_NOT_FOUND",
+       "message" : "Order does not exists.",
+}
+), 404
+    
+    order.update(set__is_deleted=True, set__del_time=datetime.utcnow())
+    
+    return jsonify({"is_deleted":bool(order.is_deleted),"message": "Order marked as deleted"}), 200
```
Comment (AkshataDudhane) [inline orders_app.py]: check why to specify bool

File: orders_app.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+@app.route('/order/<order_id>', methods=['GET']) #get order by id
+def get_order(order_id):
+    order = OrdersModel.objects(id=order_id).first()
+    
+    if not order:
+        return jsonify({"error_code":"400","message":"Order does not exists","error": "ORDER_NOT_FOUND"}), 404
+    
+    response_data = {
+        "order_id":str(order.id),
```
Comment (AkshataDudhane) [inline orders_app.py]: check str(order_id)

File: orders_app.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            return jsonify(error_response), 400
+
+    # Validate input data using Order class methods
+    order = Order(order_id=id,name=data['name'], birthday=data.get('birthday'), email=data['email'], state=data['state'], zipcode=data['zipcode']) 
+    
+    validation_result = order.validate_order()
+    if validation_result:
+        return jsonify(validation_result), 400
+    # Create a new OrdersModel instance
+    order_model = OrdersModel(
```
Comment (AkshataDudhane) [inline orders_app.py]: remove

File: orders_app.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        name=order.user.name,
+        birthday=order.user.birthday,
+        email=order.user.email,
+        state=order.user.state,
+        zipcode=order.user.zipcode
+    )
+
+    # Save the order to the database
+    order.save_order()
```
Comment (AkshataDudhane) [inline orders_app.py]: change the name to order_service

File: orders_app.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        order.update(set__is_deleted=True, set__del_time=datetime.utcnow())
+        
+        return jsonify({"is_deleted":order.is_deleted,"message": "Order marked as deleted"}), 200
+    except DoesNotExist:
+        return jsonify({"error_code": "404", "message": "Order not found", "error": "ORDER_NOT_FOUND"}), 404
+   
+    except Exception as e:
+         return jsonify({"error_code": "500", "message": "Internal Server Error", "error": str(e)}), 500
+
+@app.route('/order/<order_id>', methods=['GET']) #get order by id
```
Comment (AkshataDudhane) [inline orders_app.py]: add this in orders_service and avoid db calls from controllers

File: orders_service.py [hunk 12]
```diff
@@ -28,7 +34,103 @@ def validate_order(self):
     
     def save_order(self):
         order_model = OrdersModel(name=self.user.name, birthday=self.user.birthday, email=self.user.email, state=self.user.state, zipcode=self.user.zipcode)
-        OrdersModel.save(order_model)
+        order_model.save()
+
+    def to_dict(self):
```
Comment (shubham-zinrelo) [inline orders_service.py:34]: Why do you need this method at order_server level, you have already implemented this at ordermodel level right >

File: orders_service.py [hunk 13]
```diff
@@ -28,7 +34,103 @@ def validate_order(self):
...
+
+            order.update(set__is_deleted=True, set__del_time=datetime.utcnow())
+
+            return {"is_deleted":order.is_deleted,"message": "Order marked as deleted"}, 200
+        except DoesNotExist:
+            return {"error_code": "404", "message": "Order not found", "error": "ORDER_NOT_FOUND"}, 404
+        except Exception as e:
+            return {"error_code": "500", "message": "Internal Server Error", "error": str(e)}, 500
+    def get_order(order_id):
+        try:
```
Comment (shubham-zinrelo) [inline orders_service.py:66]: service methods are ok

**Review by shubham-zinrelo — APPROVED** [review]

**Pull request merged**